### PR TITLE
 Fix encoding object in sink with JSON schema pattern properties 

### PR DESCRIPTION
**Pull request by jedrz**

### CODE REVIEW

File: engine/lite/components/kafka-tests/src/test/scala/pl/touk/nussknacker/engine/lite/components/LiteKafkaUniversalJsonFunctionalTest.scala [hunk 1]
```diff
@@ -266,7 +271,7 @@ class LiteKafkaUniversalJsonFunctionalTest extends AnyFunSuite with Matchers wit
         schemaMapAny,
         createObjectSchemaWithPatternProperties(Map("foo_int" -> schemaInteger)),
         validationMode = Some(ValidationMode.lax)
-      ), s"#/foo_int: expected type: Integer, found: String"),
```
Comment (lciolecki) [inline engine/lite/components/kafka-tests/src/test/scala/pl/touk/nussknacker/engine/lite/components/LiteKafkaUniversalJsonFunctionalTest.scala:269]: But right now we can't see where exactly the mistake is - I mean there is a missing field name.

File: utils/json-utils/src/main/scala/pl/touk/nussknacker/engine/json/encode/BestEffortJsonSchemaEncoder.scala [hunk 2]
```diff
@@ -27,35 +27,23 @@ object BestEffortJsonSchemaEncoder {
...
-      }
-    }.sequence.map { values => Json.fromFields(values) }
-  }
+  private[encode] def encodeWithJsonValidation(value: Any, schema: Schema, fieldName: Option[String] = None): EncodeOutput =
+    encode(value, schema, fieldName).andThen { result =>
+      val json = JsonSchemaUtils.circeToJson(result)
+      schema
+        .validateData(json)
+        .leftMap(errorMsg => NonEmptyList.of(errorMsg))
+        .map(_ => result) //we return here base json (e.g. without felt defaults - there is no need to put these data as output, because json still is valid)
```
Comment (raphaelsolarski) [inline utils/json-utils/src/main/scala/pl/touk/nussknacker/engine/json/encode/BestEffortJsonSchemaEncoder.scala:36]: typo `felt` ?

File: utils/json-utils/src/main/scala/pl/touk/nussknacker/engine/json/encode/BestEffortJsonSchemaEncoder.scala [hunk 3]
```diff
@@ -81,6 +69,40 @@ object BestEffortJsonSchemaEncoder {
...
+    case ObjectField(fieldName, Some(null), Some(schema), _) if schema.isNullableSchema =>
+      Valid((fieldName, Json.Null))
+    case ObjectField(fieldName, None, Some(_), parentSchema) if parentSchema.getRequiredProperties.contains(fieldName) =>
+      error(s"Missing property: $fieldName for schema: $parentSchema.")
+    case ObjectField(fieldName, Some(value), Some(schema), _) =>
+      encode(value, schema, Some(fieldName)).map(fieldName -> _)
+  }
+
+  private def encodeFieldWithoutSchema: PartialFunction[ObjectField, WithError[(String, Json)]] = {
+    case PatternPropertySchema(schema, ObjectField(fieldName, Some(value), _, _)) =>
```
Comment (raphaelsolarski) [inline utils/json-utils/src/main/scala/pl/touk/nussknacker/engine/json/encode/BestEffortJsonSchemaEncoder.scala:92]: maybe we should left comment that now if patternProperties overlaps with properties, we chooses schema from property as schema used for encoding?

**Review by raphaelsolarski — APPROVED** [review]

**Review by lciolecki — APPROVED** [review]

File: utils/json-utils/src/main/scala/pl/touk/nussknacker/engine/json/encode/BestEffortJsonSchemaEncoder.scala [hunk 4]
```diff
@@ -131,17 +153,20 @@ object BestEffortJsonSchemaEncoder {
...
-  private[encode] def encodeWithJsonValidation(value: Any, schema: Schema, fieldName: Option[String] = None): EncodeOutput =
-    encode(value, schema, fieldName).andThen { result =>
-      val json = JsonSchemaUtils.circeToJson(result)
-      schema
-        .validateData(json)
-        .leftMap(errorMsg => NonEmptyList.of(errorMsg))
-        .map(_ => result) //we return here base json (e.g. without felt defaults - there is no need to put these data as output, because json still is valid)
+  private object PatternPropertySchema {
+    def unapply(objectField: ObjectField): Option[(Schema, ObjectField)] = {
+      for {
```
[the remark this answers is not in the document]
Comment (jedrz) [reply utils/json-utils/src/main/scala/pl/touk/nussknacker/engine/json/encode/BestEffortJsonSchemaEncoder.scala]: You're right. Map is more readable.

File: engine/lite/components/kafka-tests/src/test/scala/pl/touk/nussknacker/engine/lite/components/LiteKafkaUniversalJsonFunctionalTest.scala (hunk 1 above)
Comment (jedrz) [reply engine/lite/components/kafka-tests/src/test/scala/pl/touk/nussknacker/engine/lite/components/LiteKafkaUniversalJsonFunctionalTest.scala:269]: The original error comes from Everit validation. Now we catch such cases while encoding the value as JSON.

Comment (lciolecki) [reply engine/lite/components/kafka-tests/src/test/scala/pl/touk/nussknacker/engine/lite/components/LiteKafkaUniversalJsonFunctionalTest.scala:269]: Nice, thanks!

**Pull request merged**